scripts/annotate/set_channel_names.py

```diff
@@ -0,0 +1,127 @@
+    if args.dryrun and loglevel > 20:
+        loglevel = 20
+    logging.basicConfig(level=loglevel, format='%(levelname)s: %(message)s')
+
+    if args.dryrun:
+        logging.info("-- DRY RUN --")
+
+    target = args.target.split(':')
+    if target[0] == 'Project':
+        args.projectId = int(target[1])
+        args.screenId = None
+    elif target[0] == 'Screen':
+        args.screenId = int(target[1])
+        args.projectId = None
+
+    return args
+
+
+def fail(msg):
+    logging.critical(msg)
+    sys.exit(1)
+
+
+def read_csv(args):
+    channels = {}
+    df = pandas.read_csv(args.file, header=0, dtype=str, na_filter=False)
+    if args.projectId:
+        for index, row in df.iterrows():
+            key = f"{row['Dataset Name']}|{row['Image Name']}"
+            if key in channels:
+                fail(f"Duplicate entry in csv file: {row['Dataset Name']}, {row['Image Name']}")
+            channel_str = row[args.column]
+            channels[key] = channel_str.split(args.separator)
+    elif args.screenId:
+        for index, row in df.iterrows():
+            key = f"{row['Plate']}|{row['Well']}"
+            if key in channels:
+                fail(f"Duplicate entry in csv file: {row['Plate']}, {row['Well']}")
+            channel_str = row[args.column]
+            channels[key] = channel_str.split(args.separator)
+    return channels
+
+
+def get_image(conn, args):
+    if args.projectId:
+        project = conn.getObject("Project", args.projectId)
+        if not project:
+            fail("There's no Project with this id.")
+        for ds in project.listChildren():
+            for img in ds.listChildren():
+                yield ds, img
+    elif args.screenId:
+        screen = conn.getObject("Screen", args.screenId)
+        if not screen:
+            fail("There's no Screen with this id.")
+        for pl in screen.listChildren():
+            for well in pl.listChildren():
+                if well.getWellSample():
+                    yield pl, well
+
+
+def set_channel_name(img, channel_name, channel_index, dryrun):
+    channels = img.getChannels(noRE=True)
+    lc = channels[channel_index].getLogicalChannel()
+    if not dryrun:
+        lc.setName(channel_name)
+        lc.save()
+
+
+def main():
+    args = get_args()
+    channels = read_csv(args)
+    with cli_login() as c:
+        conn = BlitzGateway(client_obj=c.get_client())
+        for container, img in get_image(conn, args):
+            if args.projectId:
+                key = f"{container.getName()}|{img.getName()}"
+            else:
+                # img is a wellsample in that case
+                key = f"{container.getName()}|{img.getWellPos()}"
+                img = img.getImage()
+            if key not in channels:
+                logging.warning(f"No channel names found for image {key}.")
+                continue
+            img_channels = channels[key]
+            if len(img_channels) != len(img.getChannels(noRE=True)):
```

Comment:
Minor point (not a blocker): you do `img.getChannels(noRE=True)` here, then again for each channel. I think you could just do it once per image?